src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs

```diff
@@ -3,26 +3,70 @@
 
 using System;
 using System.Collections.Generic;
+using System.IO;
 using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
+using Microsoft.VisualStudio.Services.Common;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Threading;
+using NuGet.Common;
 using NuGet.Frameworks;
 using NuGet.LibraryModel;
+using NuGet.PackageManagement.VisualStudio.Exceptions;
 using NuGet.PackageManagement.VisualStudio.Utility;
 using NuGet.Packaging;
+using NuGet.Packaging.Core;
 using NuGet.ProjectManagement;
 using NuGet.ProjectManagement.Projects;
 using NuGet.ProjectModel;
+using NuGet.VisualStudio;
+using NuGet.VisualStudio.Internal.Contracts;
+using TransitiveEntry = System.Collections.Generic.IDictionary<NuGet.Frameworks.FrameworkRuntimePair, System.Collections.Generic.IList<NuGet.Packaging.PackageReference>>;
 
 namespace NuGet.PackageManagement.VisualStudio
 {
     /// <summary>
     /// An implementation of <see cref="NuGetProject"/> that interfaces with VS project APIs to coordinate
     /// packages in a package reference style project.
     /// </summary>
-    public abstract class PackageReferenceProject : BuildIntegratedNuGetProject
+    /// <remarks>Each concrete implementation is responsible of initializing <see cref="InstalledPackages"/> and <see cref="TransitivePackages"/> collections</remarks>
+    /// <typeparam name="T">A collection type for Installed and Transtive packages</typeparam>
+    /// <typeparam name="U">Type of the collection elements for Installed and Transitive packages</typeparam>
+    public abstract class PackageReferenceProject<T, U> : BuildIntegratedNuGetProject, IPackageReferenceProject where T : ICollection<U>
```

Comment:
I'm reading [your reply to Kartheek's question](https://github.com/NuGet/NuGet.Client/pull/4119/files#r805103018) (link won't work unless you "show conversation", but I don't understand why CpsPackageReferenceProject uses a different type to LegacyPackageReferenceProject.  It seems to me like it might be a leaky abstraction?  I think it's also confusing for people trying to understand the code because I would expect NuGet to have a single abstraction for what a package reference is, not one per 

On the other hand, we have to be pragmatic and refactoring one (or both) of the project systems is clearly out of scope. Additionally, if we add a new abstraction and have both CPS and Legacy project abstractions transform from their internal representations to the new abstraction, that might cause a larger number of allocations and be worse off.

So, I'm guessing there's no ideal solution here, but I just hope it's understood that I believe this change is making this class significantly more complex to understand and read. Were alternatives considered?  I really don't want to delay this PR longer, but I am worried about long term maintenance.

Reply:
Going back to @kartheekp-ms question

I refactored code up to the point I only needed to express transitive origins code only once. 

There is a way to refactor both project systems: just make the Legacy Project a special case of SDK-style Project, since Legacy Project does not support multitargeting. That requires more effort and it is not in scope of this PR.

Reply:
The alternative I considered was to not refactor both project systems, but then I got some feedback that asked for refactoring.

Reply:
Regarding the leaky abstraction, I tried to leave specific details in each concrete Project System implementation expressed in the three new abstract methods:

- GetPackageSpecAsync
- FetchInstalledPackagesList
- FetchTransitivePackagesList

Each Project System defines the way it stores its Installed and Transitive package collections. That's why `PackageReferenceProject` has type parameters.